content/router.xql

```diff
@@ -75,376 +36,300 @@ declare function router:route($jsonPaths as xs:string+, $lookup as function(xs:s
-        $router:RESPONSE_TYPE : $mediaType,
+        $router:RESPONSE_TYPE : $media-type,
         $router:RESPONSE_BODY : $body
     }
 };
 
-declare function router:match-path($config as map(*), $priority as xs:int) {
-    let $method := request:get-method() => lower-case()
-    let $path := request:get-attribute("$exist:path")
-    (: find matching route by checking each path pattern :)
-    let $routes := map:for-each($config?paths, function($pattern, $route) {
-        if (exists($route($method))) then
-            let $regex := router:create-regex($pattern)
-            return
-                if (matches($path, $regex)) then
-                    map {
-                        "path": $path,
-                        "pattern": $pattern,
-                        "config": $route($method),
-                        "regex": $regex,
-                        "spec": $config,
-                        "priority": $priority
-                    }
-                else
-                    ()
-        else
-            ()
-    })
-    return
-        if (empty($routes)) then
-            ()
+(:~
+ : resolve pointer to information in API definition
+ :
+ : @param $config the API definition
+ : @param $ref either a single string from $ref or a sequence of strings 
+ :)
+declare function router:resolve-pointer($config as map(*), $ref as xs:string*) {
+    let $parts := 
+        if (starts-with($ref, "#/")) then
+            tokenize($ref, "/") (: this is a $ref in the document :)
         else
-            $routes
+            $ref (: internal spec lookup :)
+
+    return router:resolve-ref($config, $parts)
 };
 
-declare function router:process($routes as map(*)*, $lookup as function(*)) {
-    (: if there are multiple matches, prefer the one matching the longest pattern :)
-    let $route := sort($routes, (), function($route) {
-            string-length(replace($route?pattern, "\{[^\}]+\}", "?")) + $route?priority
-        }) => reverse() => head()
-    let $loginDomain := router:login-domain($route?spec)
-    let $parameters := map:merge((
-        router:map-request-parameters($route?config),
-        router:map-path-parameters($route, $route?path)
-    ))
-    let $info := $route?spec?info
-    let $request := map {
-        "parameters": $parameters,
-        "body": router:request-body($route?config),
-        "loginDomain": $loginDomain,
-        "info": $info,
-        "config": $route
+(:~
+ : Maps a request to the configured handler 
+ : Loads API definitions, matches routes
+ : Calls route handler function returned by $lookup function
+ :
+ : Route specificity rules:
+ : 1. normalize patterns: replace placeholders with "?"
+ : 2. use the matching route with the longest normalized pattern
+ : 3. If two paths have the same (normalized) length, prioritize by appearance in API files, first one wins
+ :)
+declare function router:route($api-files as xs:string+, $lookup as function(xs:string) as function(*)?, $middlewares as function(*)*) {
```

Comment:
I am open for other name suggestions. I chose simple because it is ... well a simpler version. As for the separate module: I tried hard to not have to do this but I encountered an issue with eXist-db's handling of circular dependencies. module `A` cannot depend on module `B` that depends on `A`.
router->auth->router would be a problem then.

Reply:
The dependency graph is actually `auth->rutil->router` and `auth` is needed for the simple router.
There is the possibility to put both routing functions into one module which then include all the dependencies.
That would mean that most of the functions currently in the router module must be moved to one with a different name.
Would that be a better solution?

Reply:
In addition to that, how about moving getDBUser into the auth module?

Reply:
Mmmmh, I cannot reproduce the circular dependency problem. Adding a two argument version of `router:route` works fine for me even if I import `auth.xql`. Let's discuss in a chat later maybe.

Reply:
Oh, OK. Well then, I stand corrected. Did you check that on existdb v5.2.0? 
The issue is definitely there but might be limited to direct cyclical dependencies, where `A` imports `B` and vice versa. I was a little too cautious then, I guess.
I am still in favor of having the router module to consist of only the functions that are important to the developers using it.

So definitely

- router:response#2 and router:response#3 
- router:route#2 and router:route#3
- router:match-route, router:process-request and router:execute-handler (even though private do help to understand the inner workings)

Other functions might be moved to a new module to narrow down the focus of a developer trying to grasp what is going on inside the router.
I am not sure if router:resolve-pointer would be better moved to rutil module.
What do you think?

Reply:
There is an issue with cyclical dependencies, but it occurs only when importing **two** library modules which depend on each other in a main module.

Because of that, the structure changed: a facade module "roaster" now is the main entry point for users of this library.